### PR TITLE
MXPredReshape bug: need to reshape softmax_label

### PR DESCRIPTION
**Pull request by chsin**
## Description ##

The fix in #11493 allows reshape only for networks that don't use SoftmaxOutput, so none of the pretrained image classification networks, e.g. inception, can be reshaped. The current test did not catch this because it only used gluon blocks. The reason SoftmaxOutput is weird is because it requires the label "softmax_label" to be reshaped even though "softmax_label" is only used as an input for training and should not even be looked at for prediction, [which has been mentioned in other context before](https://stackoverflow.com/questions/44947104/mxnet-label-shapes-dont-match-names-specified-by-label-names). I don't know what's the best way to deal with SoftmaxOutput's label reshape requirement because I can't find a way to get the label names. I am sending this pull request with an edit that allows resnet and inception to be reshaped, along with a test that covers this edge case, but I hope someone can comment on a more robust solution.

I found this issue when I made the following edits to [image-classification-predict.cc:223](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/predict-cpp/image-classification-predict.cc#L223). Before this change, the following would print out -1 and I traced it to `c_predict_api.cc:304` where `newShape.Size() != arr.shape().Size()` for when `arg_names[I]` is `softmax_label`: 
```c++
  // Create Predictor
  PredictorHandle old_hnd = nullptr;
  const mx_uint old_input_shape_data[4] = { 2,
                                        static_cast<mx_uint>(channels),
                                        static_cast<mx_uint>(height),
                                        static_cast<mx_uint>(width) };
  MXPredCreate(static_cast<const char*>(json_data.GetBuffer()),
               static_cast<const char*>(param_data.GetBuffer()),
               static_cast<int>(param_data.GetLength()),
               dev_type,
               dev_id,
               num_input_nodes,
               input_keys,
               input_shape_indptr,
               old_input_shape_data,
               &old_hnd);
  assert(old_hnd);

  int e = MXPredReshape(num_input_nodes,
                        input_keys,
                        input_shape_indptr,
                        input_shape_data,
                        old_hnd,
                        &pred_hnd);
  printf("%d\n", e);
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by chinakook** [comment]
In the inference stage, it's no need to use labels or SoftmaxOutput.

**Comment by chinakook** [comment]
You can write a script to deploy the symbol by replace SoftmaxOutput with SoftmaxActivate.

**Comment by chsin** [comment]
@chinakook, To predict using the C API, you use `MXPredCreate` to load the pretrained symbols and weights and this works fine for [all the mxnet pretrained models that use SoftmaxOutput](http://data.mxnet.io/models/imagenet/), but the problem is that `MXPredReshape` doesn't work even though the models that were loaded properly. This causes reshape to fail for all the pretrained mxnet models (almost all use SoftmaxOutput) even though predcreate is fine. This is not an acceptable situation. I don't quite understand your comment, but you seem to be suggesting that this inconsistency is acceptable coding practice and it is the user that needs to clean up this mess by writing a script?

**Comment by lupesko** [comment]
Thanks for identifying the issue and filing this PR @chsin .
Looping in additional folks to review: @apeforest @anirudh2290 @eric-haibin-lin

**Comment by stu1130** [comment]
bounce again @apeforest @anirudh2290 @eric-haibin-lin

**Comment by vandanavk** [comment]
@apeforest @anirudh2290 @eric-haibin-lin ping for review

**Comment by vandanavk** [comment]
@mxnet-label-bot [pr-awaiting-review]

File: src/c_api/c_predict_api.cc [hunk 1]
```diff
@@ -304,7 +304,8 @@ int MXPredReshape(mx_uint num_input_nodes,
   for (size_t i=0; i < arg_names.size(); ++i) {
     TShape newShape = arg_shapes[i];
     NDArray &arr = p->arg_arrays[i];
-    if (new_shape.count(arg_names[i]) != 0) {
+    if (new_shape.count(arg_names[i]) != 0 ||
+        strcmp("softmax_label", arg_names[i].c_str()) == 0) {
```
Comment (apeforest) [inline src/c_api/c_predict_api.cc]: Still trying to understand the problem. But this hack is not acceptable.

**Comment by Roshrini** [comment]
@chsin Can you please address review comments? Thanks!

**Comment by ankkhedia** [comment]
@chsin ping! Could you please address the comments? Thanks!

**Comment by kalyc** [comment]
@chsin Thanks for your contribution! We would like to take the process of merging this PR forward - could you please address comments above?

**Comment by stu1130** [comment]
@mxnet-label-bot update [pr-awaiting-response]
@chsin let us know if you need any help!

**Comment by vandanavk** [comment]
adding @vikas89 for review

**Comment by roywei** [comment]
@chsin  Thanks for the contribution, looks like our CI hanged, could trigger it by push an empty commit?

**Comment by sandeep-krishnamurthy** [comment]
@chsin - Can you please rebase and retrigger CI?

**Comment by stu1130** [comment]
@anirudh2290 ping for review @chsin could you rebase and retrigger CI thanks!

**Comment by sandeep-krishnamurthy** [comment]
@chsin - Thanks for your contributions. Like @apeforest mentioned, this change looks like not generalizable. What happens if name of last layer is not 'softmax_label' for example in Gluon case it will may be prefixed with name_scope.
